cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,151 @@
+# CFP-30984: toFQDN DNS proxy HA
+
+**SIG: SIG-POLICY**
+
+**Begin Design Discussion:** 2024-04-08
+
+**Cilium Release:** 1.17
+
+**Authors:** Hemanth Malla <hemanth.malla@datadoghq.com>
+
+## Summary
+
+Cilium agent uses a proxy to intercept all DNS queries and obtain necessary information for enforcing toFQDN network policies. However, the lifecycle of this proxy is coupled with cilium agent. When an endpoint has a toFQDN network policy in place, cilium installs a redirect to capture all DNS traffic. So, when the agent is unavailable all DNS requests time out, including when DNS name to IP address mappings are already in place for this name.DNS policy unload on shutdown can be enabled on agent, but it works only when L7 policy is set to * and agent is shutdown gracefully.
+
+This CFP introduces a standalone DNS proxy that can be run alongside cilium agent which should eliminate the hard dependency on cilium agent atleast for names that are already resolved. 
+
+## Motivation
+
+Users rely on toFQDN policies to enforce network policies against traffic to remote destinations outside the cluster, typically to blob storage / other services on the internet. Rolling out cilium agent should not result in traffic being dropped. Introducing a high availablility (HA) mode will allow for adoption of toFQDN network policies in critical environments.
+
+## Goals
+
+* Introduce a streaming gRPC API for exchanging FQDN related policy information and endpoint related metadata.
+* Introduce standalone DNS proxy (SDP) that binds on the same port as built-in proxy with SO_REUSEPORT and uses the above mentioned API to notify agent of new DNS resolutions.
```

Comment:
This is for redirecting packets either to SDP or CA dnsproxy. For updating dns->ip mappings it uses grpc channel and there will be retry on failures i guess

Reply:
Did this part get some more thought? I don't think it necessarily has to be in scope to solve this problem in the initial CFP, but I can see you're already thinking about this. It would be nice to at least clarify what the intended behavior is right now even if a better solution is planned for later. This is also what I'm thinking about from this thread below: https://github.com/cilium/design-cfps/pull/32/files/616ae893539fcab4a47e15de023215ddae46eec9#r1710516516 .